pdns/opensslsigners.cc

```diff
@@ -910,6 +913,239 @@ void OpenSSLECDSADNSCryptoKeyEngine::fromPublicKeyString(const std::string& inpu
 }
 #endif
 
+#ifdef HAVE_LIBCRYPTO_EDDSA
+class OpenSSLEDDSADNSCryptoKeyEngine : public DNSCryptoKeyEngine
+{
+public:
+  explicit OpenSSLEDDSADNSCryptoKeyEngine(unsigned int algo) : DNSCryptoKeyEngine(algo)
+  {
+
+    int ret = RAND_status();
+    if (ret != 1) {
+      throw runtime_error(getName()+" insufficient entropy");
+    }
+
+    d_edkey = EVP_PKEY_new();
+    if (d_edkey == nullptr) {
+      throw runtime_error(getName()+" allocation of key structure failed");
+    }
+
+#ifdef HAVE_LIBCRYPTO_ED25519
+    if(d_algorithm == 15) {
+      d_len = 32;
+      d_id = NID_ED25519;
+    }
+#endif
+#ifdef HAVE_LIBCRYPTO_ED448
+    if (d_algorithm == 16) {
+      d_len = 57;
+      d_id = NID_ED448;
+    }
+#endif
+    if (d_len == 0) {
+      EVP_PKEY_free(d_edkey);
+      throw runtime_error(getName()+" unknown algorithm "+std::to_string(d_algorithm));
+    }
+  }
+
+  ~OpenSSLEDDSADNSCryptoKeyEngine()
+  {
+    EVP_PKEY_free(d_edkey);
+  }
+
+  string getName() const override { return "OpenSSL EDDSA"; }
+  int getBits() const override { return d_len; }
+
+  void create(unsigned int bits) override;
+  storvector_t convertToISCVector() const override;
+  std::string sign(const std::string& hash) const override;
+  bool verify(const std::string& msg, const std::string& signature) const override;
+  std::string getPubKeyHash() const override;
+  std::string getPublicKeyString() const override;
+  void fromISCMap(DNSKEYRecordContent& drc, std::map<std::string, std::string>& stormap) override;
+  void fromPublicKeyString(const std::string& content) override;
+  bool checkKey() const override;
+
+  static std::shared_ptr<DNSCryptoKeyEngine> maker(unsigned int algorithm)
+  {
+    return std::make_shared<OpenSSLEDDSADNSCryptoKeyEngine>(algorithm);
+  }
+
+private:
+  size_t d_len{0};
+  int d_id{0};
+
+  EVP_PKEY *d_edkey = nullptr;
+};
+
+bool OpenSSLEDDSADNSCryptoKeyEngine::checkKey() const
+{
+  return (d_edkey != nullptr);
+}
+
+void OpenSSLEDDSADNSCryptoKeyEngine::create(unsigned int bits)
+{
+  auto pctx = EVP_PKEY_CTX_new(d_edkey, nullptr);
```

Comment:
I think we should generate the context from the algorithm id:
```
auto pctx = EVP_PKEY_CTX_new_id(d_id, nullptr);
```

Otherwise I don't see how the correct type of key can be generated.